lab4.md

```diff
@@ -158,6 +158,67 @@ In Python code you'd type in `os.getenv("Http_X_Output_Mode")`.
 
 You can see that all other HTTP context is also provided such as `Content-Length` when the `Http_Method` is a `POST`, the `User_Agent`, Cookies and anything else you'd expect to see from a HTTP request.
 
+### Making use of logging
+
+The way OpenFaaS watchdog works is that it passses in the HTTP request and reads a HTTP response via the standard I/O streams `stdin` and `stdout` accordingly.
+This means that your process does not need to know anything about the web or HTTP.
+
+An interesting case is when it finishes with a non-zero exit code and `stderr` is not empty.
+By default function `stdout/stderr` is combined and `stderr` is not printed to the logs.
+
+Lets check that with the `hello-openfaas` function from Lab 3.
+
+Change the `handler.py` code to
+
+```python
+import sys
+import json
+
+def handle(req):
+
+    sys.stderr.write("This should be an error message.\n")
+    return json.dumps({"hello": "world"})
+```
+
+Now invoke the function with
+
+```bash
+echo | faas invoke hello-openfaas
+```
+
+You should see the combined output:
+
+```
+This should be an error message.
+Hello World!
```

Comment:
This output is not correct. Please update your function and run it again. `json.dumps` would produce a JSON string.